src/waiter.c

```diff
@@ -0,0 +1,311 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <stdlib.h>
+#include <stdalign.h>
+#include <stdbool.h>
+#include <inttypes.h>
+
+#include "macro_utils/macro_utils.h"
+
+#include "c_logging/xlogging.h"
+
+#include "c_pal/thandle.h"
+#include "c_pal/sm.h"
+#include "c_pal/srw_lock.h"
+
+#include "c_util/async_op.h"
+#include "c_util/rc_ptr.h"
+
+#include "c_util/waiter.h"
+
+typedef struct WAITER_TAG
+{
+    SM_HANDLE sm;
+    SRW_LOCK_HANDLE lock;
+    THANDLE(RC_PTR) data;
+    NOTIFICATION_CALLBACK notification_callback;
+    void* notification_callback_context;
+    NOTIFY_COMPLETE_CALLBACK notify_complete_callback;
+    void* notify_complete_callback_context;
+    bool registered;
```

Comment:
>registered [](http://example.com/codeflow?start=9&length=10)

some comments would be nice here to say what's the meaning of these variables.

Reply:
Done